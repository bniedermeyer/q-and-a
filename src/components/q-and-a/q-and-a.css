:host {
  display: flex;
  flex-direction: column;
  padding: 10px;
  font-family: canada-type-gibson, sans-serif;
}
