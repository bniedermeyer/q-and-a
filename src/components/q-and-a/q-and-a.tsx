import { Component, Prop, Watch, Host, h } from '@stencil/core';
import state from '../../store';

@Component({
  tag: 'q-and-a',
  styleUrl: 'q-and-a.css',
  shadow: true,
})
export class QAndA {
  /**
   * The endpoint that questions/upvotes will be posted to
   */
  @Prop() askEndpoint: string;
  /**
   * The endpoint the list of questions will be retrieved from
   */
  @Prop() retrieveEndpoint: string;
  /**
   * The optional id of the user asking a question. Will be sent with the question if present.
   */
  @Prop() userId: string;
  /**
   * The correlation id of the q-and-a session. If present, will be sent with questions and used to retrieve questions by appending a `correlationId` query parameter to the `retrieveEndpoint` url.
   */
  @Prop() correlationId: string;
  /**
   * Primary color. Used mainly for button borders. Defaults to #10915b
   */
  @Prop() primaryColor: string;
  /**
   * Secondary color. Used for question and header text color. Defaults to  #112378
   */
  @Prop() secondaryColor: string;

  connectedCallback() {
    if (this.askEndpoint) {
      state.askEndpoint = this.askEndpoint;
    }
    if (this.retrieveEndpoint) {
      state.retrieveEndpoint = this.retrieveEndpoint;
    }
    if (this.userId) {
      state.userId = this.userId;
    }
    if (this.correlationId) {
      state.correlationId = this.correlationId;
    }
    if (this.primaryColor) {
      state.primaryColor = this.primaryColor;
    }
    if (this.secondaryColor) {
      state.secondaryColor = this.secondaryColor;
    }
  }

  @Watch('correlationId')
  updateCorrelationId(newId: string, oldId: string) {
    console.log(`Q&A Correlation ID was updated from ${oldId} to ${newId}`);
    state.correlationId = newId;
  }

  render() {
    return (
      <Host>
        <qa-question-form></qa-question-form>
      </Host>
    );
  }
}
