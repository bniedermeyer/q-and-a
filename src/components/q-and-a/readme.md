# q-and-a

Please note that while the component does a great job submitting and fetching questions from the audience, it is up to you to implement the logic in the backend. The properties, and the options they configure, are useful for configuring the behavior of the component and what information it sends to your backend.

<!-- Auto Generated Below -->


## Properties

| Property              | Attribute               | Description                                                                                                                                                                                                                                                                                                                                                                                                                     | Type      | Default                   |
| --------------------- | ----------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | --------- | ------------------------- |
| `askEndpoint`         | `ask-endpoint`          | The endpoint that questions/upvotes will be posted to.                                                                                                                                                                                                                                                                                                                                                                          | `string`  | `'/ask'`                  |
| `correlationId`       | `correlation-id`        | The correlation id of the q-and-a session. If present, will be sent with questions and used to retrieve questions by appending a `correlationId` query parameter to the `retrieveEndpoint` url. Changes to this property are tracked and updated within the component. A good use case for this property is to track which talk is being given at a time. This way you can only display questions relevant to the current talk. | `string`  | `undefined`               |
| `firebaseAskFn`       | `firebase-ask-fn`       | Use to customize the name of your firebase callable function to submit questions. Default is `askQuestion`                                                                                                                                                                                                                                                                                                                      | `string`  | `'askQuestion'`           |
| `firebaseAuthDomain`  | `firebase-auth-domain`  | The auth domain for your firebase project. Needed for Safari compatibility. You probably don't need to adjust this unless you have issues with your app's authentication in general.                                                                                                                                                                                                                                            | `string`  | `'noop'`                  |
| `firebaseDb`          | `firebase-db`           | URL for the Firebase Realtime Database. Required if `useFirebase` = true                                                                                                                                                                                                                                                                                                                                                        | `string`  | `undefined`               |
| `firebaseIncrementFn` | `firebase-increment-fn` | Use to customize the name of your firebase callable function to increment question counts. Default is `incrementQuestion`                                                                                                                                                                                                                                                                                                       | `string`  | `'incrementQuestion'`     |
| `firebaseProjectId`   | `firebase-project-id`   | The project ID for your firebase project. Required if `useFirebase` = true                                                                                                                                                                                                                                                                                                                                                      | `string`  | `undefined`               |
| `firebaseToken`       | `firebase-token`        | Web API Key for your Firebase project. Required if `useFirebase` = true                                                                                                                                                                                                                                                                                                                                                         | `string`  | `undefined`               |
| `incrementEndpoint`   | `increment-endpoint`    | The endpoint to post increment commands to.                                                                                                                                                                                                                                                                                                                                                                                     | `string`  | `'/ask'`                  |
| `placeholder`         | `placeholder`           | Controls whether placholder text is displayed.                                                                                                                                                                                                                                                                                                                                                                                  | `boolean` | `false`                   |
| `placeholderText`     | `placeholder-text`      | Text to display if `placeholder` = true                                                                                                                                                                                                                                                                                                                                                                                         | `string`  | `'Q&A will open shortly'` |
| `pollingInterval`     | `polling-interval`      | The interval in which the questions should be fetched in ms. Defaults to 10000ms (10 seconds).                                                                                                                                                                                                                                                                                                                                  | `number`  | `10000`                   |
| `primaryColor`        | `primary-color`         | Primary color. Used mainly for button borders.                                                                                                                                                                                                                                                                                                                                                                                  | `string`  | `'#007745 '`              |
| `retrieveEndpoint`    | `retrieve-endpoint`     | The endpoint the list of questions will be retrieved from.                                                                                                                                                                                                                                                                                                                                                                      | `string`  | `'/questions'`            |
| `secondaryColor`      | `secondary-color`       | Secondary color. Used for question and header text color.                                                                                                                                                                                                                                                                                                                                                                       | `string`  | `'#112378'`               |
| `useFirebase`         | `use-firebase`          | Whether to use REST polling or Firebase                                                                                                                                                                                                                                                                                                                                                                                         | `boolean` | `false`                   |
| `userId`              | `user-id`               | The optional id of the user asking a question. Will be sent with the question if present. Cannot be changed once initialized.                                                                                                                                                                                                                                                                                                   | `string`  | `undefined`               |


## Shadow Parts

| Part                 | Description |
| -------------------- | ----------- |
| `"placeholder-text"` |             |


## Dependencies

### Depends on

- [qa-question-form](../qa-question-form)

### Graph
```mermaid
graph TD;
  q-and-a --> qa-question-form
  style q-and-a fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
