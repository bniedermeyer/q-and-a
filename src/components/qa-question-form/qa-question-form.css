:host {
  display: block;
}

.qa-post-submit-message {
  font-size: 0.75rem;
  padding: 10px;
}

#qa-question-error {
  color: red;
}

#qa-submit-question-btn {
  height: 30px;
  font-weight: bold;
  background-color: #fff;
}
