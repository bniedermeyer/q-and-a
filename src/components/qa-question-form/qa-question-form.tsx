import { Component, State, h } from '@stencil/core';
import { askQuestion } from '../../utils/data-fetching-utils';
import state from '../../store';

@Component({
  tag: 'qa-question-form',
  styleUrl: 'qa-question-form.css',
  shadow: true,
})
export class QaQuestionForm {
  /**
   * The content of the question to be submitted
   */
  @State() question: string;
  /**
   * Whether to display a submit confirmation
   */
  @State() displayConfirmation: boolean;
  /**
   * Error messages that are created while submitting a question
   */
  @State() error: string;

  async handleSubmit(e: Event): Promise<void> {
    e.preventDefault();
    if (!this.question) {
      this.error = 'Please type a question before submitting!';
      this.displayConfirmation = false;
      return;
    }
    try {
      await askQuestion(this.question);
      this.displayConfirmation = true;
    } catch (error) {
      this.error = error.message;
    }
    this.question = '';
  }

  private handleQuestionChange = e => {
    this.question = e.target.value;
    if (this.displayConfirmation) {
      this.displayConfirmation = false;
    }
    if (this.error) {
      this.error = undefined;
    }
  };

  render() {
    return (
      <form id="qa-question-form" onSubmit={e => this.handleSubmit(e)} part="form">
        <div>
          <textarea
            name="question"
            id="question"
            cols={30}
            rows={3}
            value={this.question}
            onChange={this.handleQuestionChange}
            aria-label="Submit your question here"
            part="question-textarea"
          ></textarea>
        </div>
        <button type="submit" id="qa-submit-question-btn" style={{ color: state.primaryColor, border: `3px solid ${state.primaryColor}` }} part="submit-button">
          Ask Question
        </button>
        {this.displayConfirmation ? (
          <span class="qa-post-submit-message" id="qa-question-confirm" style={{ color: state.secondaryColor }} part="confirmation-message">
            Question Asked!
          </span>
        ) : null}
        {this.error ? (
          <span class="qa-post-submit-message" id="qa-question-error" part="error-message">
            {this.error}
          </span>
        ) : null}
      </form>
    );
  }
}
