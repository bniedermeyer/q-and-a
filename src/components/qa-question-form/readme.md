# qa-question-form



<!-- Auto Generated Below -->


## Shadow Parts

| Part                     | Description |
| ------------------------ | ----------- |
| `"confirmation-message"` |             |
| `"error-message"`        |             |
| `"form"`                 |             |
| `"loading-message"`      |             |
| `"question-textarea"`    |             |
| `"submit-button"`        |             |


## Dependencies

### Used by

 - [q-and-a](../q-and-a)

### Graph
```mermaid
graph TD;
  q-and-a --> qa-question-form
  style qa-question-form fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
