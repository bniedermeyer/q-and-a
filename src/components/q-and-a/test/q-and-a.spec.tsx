jest.mock('../../../utils/utils');

import { newSpecPage } from '@stencil/core/testing';
import { QAndA } from '../q-and-a';
import state, { disposeStore } from '../../../store';
import { QaQuestionForm } from '../../qa-question-form/qa-question-form';

describe('q-and-a', () => {
  beforeEach(() => {
    disposeStore();
  });

  it('renders', async () => {
    const page = await newSpecPage({
      components: [QAndA, QaQuestionForm],
      html: `<q-and-a></q-and-a>`,
    });
    expect(page.root).toMatchSnapshot();
  });

  it('should set the state according to the properties supplied', async () => {
    const page = await newSpecPage({
      components: [QAndA, QaQuestionForm],
      html: `<q-and-a ask-endpoint="/test-endpoint"
                      retrieve-endpoint="/test-retrieve"
                      user-id="abc123"
                      correlation-id="corr123"
                      primary-color="#585859"
                      secondary-color="#72d373"
                      polling-interval="3000" ></q-and-a>`,
    });
    expect(page.root).toMatchSnapshot();
    expect(state).toMatchSnapshot();
  });

  it('should update the correlation id if it changes', async () => {
    const page = await newSpecPage({
      components: [QAndA, QaQuestionForm],
      html: `<q-and-a></q-and-a>`,
    });
    expect(state).toMatchSnapshot();
    page.rootInstance.correlationId = 'corr1234';
    await page.waitForChanges();
    expect(state).toMatchSnapshot();
  });
});
