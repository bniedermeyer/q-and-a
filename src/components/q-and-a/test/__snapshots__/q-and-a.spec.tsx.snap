// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`q-and-a renders 1`] = `
<q-and-a>
  <mock:shadow-root>
    <qa-question-form>
      <mock:shadow-root>
        <form id="qa-question-form" part="form">
          <div><textarea aria-label="Submit your question here" cols="30" id="question" name="question" part="question-textarea" rows="3"></textarea>
          </div>
          <button id="qa-submit-question-btn" part="submit-button" type="submit" style="color: #007745 ; border: 3px solid #007745 ;">
            Ask Question
          </button>
        </form>
      </mock:shadow-root>
    </qa-question-form>
  </mock:shadow-root>
</q-and-a>
`;

exports[`q-and-a should set the state according to the properties supplied 1`] = `
<q-and-a ask-endpoint="/test-endpoint" correlation-id="corr123" polling-interval="3000" primary-color="#585859" retrieve-endpoint="/test-retrieve" secondary-color="#72d373" user-id="abc123">
  <mock:shadow-root>
    <qa-question-form>
      <mock:shadow-root>
        <form id="qa-question-form" part="form">
          <div><textarea aria-label="Submit your question here" cols="30" id="question" name="question" part="question-textarea" rows="3"></textarea>
          </div>
          <button id="qa-submit-question-btn" part="submit-button" type="submit" style="color: #585859; border: 3px solid #585859;">
            Ask Question
          </button>
        </form>
      </mock:shadow-root>
    </qa-question-form>
  </mock:shadow-root>
</q-and-a>
`;

exports[`q-and-a should set the state according to the properties supplied 2`] = `
Object {
  "askEndpoint": "/test-endpoint",
  "correlationId": "corr123",
  "incrementEndpoint": "/ask",
  "primaryColor": "#585859",
  "questions": Array [],
  "retrieveEndpoint": "/test-retrieve",
  "secondaryColor": "#72d373",
  "userId": "abc123",
}
`;

exports[`q-and-a should update the correlation id if it changes 1`] = `
Object {
  "askEndpoint": "/ask",
  "correlationId": "",
  "incrementEndpoint": "/ask",
  "primaryColor": "#007745 ",
  "questions": Array [],
  "retrieveEndpoint": "/questions",
  "secondaryColor": "#112378",
  "userId": "",
}
`;

exports[`q-and-a should update the correlation id if it changes 2`] = `
Object {
  "askEndpoint": "/ask",
  "correlationId": "corr1234",
  "incrementEndpoint": "/ask",
  "primaryColor": "#007745 ",
  "questions": Array [],
  "retrieveEndpoint": "/questions",
  "secondaryColor": "#112378",
  "userId": "",
}
`;
